bootstrap/src/basis/file.ml

```diff
@@ -201,38 +198,67 @@ let read_into_hlt buffer t =
   let _ = Read.(submit ~buffer t |> complete_hlt) in
   ()
 
-(* external write_inner: _?bytes -> t >os-> sint *)
-external write_inner: Stdlib.Bytes.t -> t -> sint = "hm_basis_file_write_inner"
-
-let rec write buffer t =
-  match Bytes.Cursor.index (Bytes.Slice.base buffer) < Bytes.Cursor.index (Bytes.Slice.past buffer)
-  with
-  | false -> None
-  | true -> begin
-      let bytes = bytes_of_slice buffer in
-      let value = write_inner bytes t in
-      match Sint.(value < kv 0L) with
-      | true -> Some (Error.of_value value)
-      | false ->
-        write (
-          Bytes.Slice.of_cursors ~base:(Bytes.Cursor.seek value (Bytes.Slice.base buffer))
-            ~past:(Bytes.Slice.base buffer)
-        ) t
-    end
+module Write = struct
+  type file = t
+  type inner = uns
+  type t = {
+    inner: inner;
+    buffer: Bytes.Slice.t;
+  }
+
+  external submit_inner: Stdlib.Bytes.t -> file -> inner = "hm_basis_file_write_submit_inner"
+
+  let submit buffer file =
+    let bytes = bytes_of_slice buffer in
+    let inner = finalize_user_data (submit_inner bytes file) in
+    { inner; buffer }
+
+  let complete t =
+    let value = complete_inner t.inner in
+    match Sint.(value < kv 0L) with
+    | true -> Error (Error.of_neg_errno value)
+    | false -> begin
+        let base = Bytes.(Slice.base t.buffer |> Cursor.seek (Uns.bits_of_sint value)) in
+        let past = Bytes.Slice.past t.buffer in
+        let buffer = Bytes.Slice.of_cursors ~base ~past in
+        Ok buffer
+      end
+
+  let complete_hlt t =
+    match complete t with
+    | Ok buffer -> buffer
+    | Error error -> halt (Error.to_string error)
+
+end
+
+let write buffer t =
+  let rec f buffer t = begin
+    match Bytes.Slice.length buffer = 0L with
```

Comment:
Maybe put an `assert` in the C layer that length is 0.

Reply:
Do you mean put it there so that we don't have to recursively call `write` here? As far as I can tell from reading `man 2 write` `RETURN VALUE` section, partial writes shouldn't be considered an error.

Reply:
No, I mean that this code assures the C layer never gets used for 0-length writes, and that we should `assert` in the C code that the write length is non-zero. That documents expectations in the C layer, so that if we ever refactor in a way that makes this no longer true we are at least made aware enough to add extra tests.